File: vbnet/ExampleEnumerate.vb
Imports Tinkerforge

Module ExampleEnumerate
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223

    Sub EnumerateCB(ByVal sender As IPConnection, _
                    ByVal uid As String, _
                    ByVal connectedUid As String, _
                    ByVal position As Char, _
                    ByVal hardwareVersion() As Short, _
                    ByVal firmwareVersion() As Short, _
                    ByVal deviceIdentifier As Integer, _
                    ByVal enumerationType As Short)
        System.Console.WriteLine("UID:               {0}", uid)
        System.Console.WriteLine("Enumeration Type:  {0}", enumerationType)

        If enumerationType = IPConnection.ENUMERATION_TYPE_DISCONNECTED Then
            System.Console.WriteLine("")
            Return
        End If

        System.Console.WriteLine("Connected UID:     {0}", connectedUid)
        System.Console.WriteLine("Position:          {0}", position)
        System.Console.WriteLine("Hardware Version:  {0}.{1}.{2}", _
                                 hardwareVersion(0), hardwareVersion(1), hardwareVersion(2))
        System.Console.WriteLine("Firmware Version:  {0}.{1}.{2}", _
                                 firmwareVersion(0), firmwareVersion(1), firmwareVersion(2))
        System.Console.WriteLine("Device Identifier: {0}", deviceIdentifier)
        System.Console.WriteLine("")
    End Sub

    Sub Main()
        ' Create connection and connect to brickd
        Dim ipcon As New IPConnection()
        ipcon.Connect(HOST, PORT)

        ' Register Enumerate Callback
        AddHandler ipcon.EnumerateCallback, AddressOf EnumerateCB

        ' Trigger Enumerate
        ipcon.Enumerate()

        System.Console.WriteLine("Press key to exit")
        System.Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module

File: delphi/ExampleEnumerate.pas
program ExampleEnumerate;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, Device;

type
  TExample = class
  private
    ipcon: TIPConnection;
  public
    procedure EnumerateCB(sender: TIPConnection;
                          const uid: string; const connectedUid: string; const position: char;
                          const hardwareVersion: TVersionNumber; const firmwareVersion: TVersionNumber;
                          const deviceIdentifier: word; const enumerationType: byte);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;

var
  e: TExample;

{ Print incoming enumeration }
procedure TExample.EnumerateCB(sender: TIPConnection;
                               const uid: string; const connectedUid: string; const position: char;
                               const hardwareVersion: TVersionNumber; const firmwareVersion: TVersionNumber;
                               const deviceIdentifier: word; const enumerationType: byte);
begin
  WriteLn('UID:               ' + uid);
  WriteLn('Enumerate Type:    ' + IntToStr(enumerationType));

  if (enumerationType <> IPCON_ENUMERATION_TYPE_DISCONNECTED) then begin
    WriteLn('Connected UID:     ' + connectedUid);
    WriteLn('Position:          ' + position);
    WriteLn('Hardware Version:  ' + IntToStr(hardwareVersion[0]) + '.' +
                                    IntToStr(hardwareVersion[1]) + '.' +
                                    IntToStr(hardwareVersion[2]));
    WriteLn('Firmware Version:  ' + IntToStr(firmwareVersion[0]) + '.' +
                                    IntToStr(firmwareVersion[1]) + '.' +
                                    IntToStr(firmwareVersion[2]));
    WriteLn('Device Identifier: ' + IntToStr(deviceIdentifier));
  end;

  WriteLn('');
end;

procedure TExample.Execute;
begin
  { Create connection and connect to brickd }
  ipcon := TIPConnection.Create;
  ipcon.Connect(HOST, PORT);

  { Register enumeration callback to "EnumerateCB }
  ipcon.OnEnumerate := {$ifdef FPC}@{$endif}EnumerateCB;

  ipcon.Enumerate;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.
